feat: Native ELFLoader support

nderjung opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This PR introduces native [ELFLoader](https://github.com/unikraft/app-elfloader) support into KraftKit which in turn enables running arbitrary statically linked Linux Userspace binaries via the `kraft run` subcommand.

To accomplish this, a number of changes were necessary which are incorporated in this PR:

- The OCI PackageManager was reworked slightly to accommodate for use of ephemeral registries so as to fetch sources from the new ELFLoader registry which is held at `loaders.unikraft.org`;
- Accommodation for the initram filesystem in the Machine Interface was adjusted in the status so that the interpration of the Linux Userspace binary can be delivered to the ELFLoader unikernel application;
- The `kraft run` subcommand was adjusted to accommodate for both determining and preparing the Machine Interface for instantiation based on different input positional arguments. By consequence, four entrypoints are now are possible:
  
| Runner | Purpose |
|-|-|
| `linuxu` | Determination of a Linux Userspace binary and actualization of the ELFLoader unikernel application (new) |
| `kernel` | Determination of a Unikraft kernel and instantiation of the unikernel (existing) |
| `project` | Determination of the positional argument (or therelackof and thus the current working directory) representing a project directory, one which contains a `kraft.yaml`/`Kraftfile` such that the user simply can run `kraft run` to instantiate their current unikernel (existing) |
| `package` | Determination of the positional argument representing a package such that the user can retrieve and instantiate a unikernel via the compatible package manager |

At a high-level, the user simply needs to provide a positional argument of a Linux userspace application to realise the functionality provided in this PR, for example:

```bash
git clone https://github.com/unikraft/app-elfloader.git;
cd app-elfloader/examples/helloworld;
make;
kraft run helloworld_static;
```

nderjung committed:
feat(manifest): Convert `manifestManager` into pointer

This allows the structure to be manipulated from within internal
methods, allowing for structure field values.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(oci): Convert `ociManager` into pointer

This allows the structure to be manipulated from within internal
methods, allowing for structure field values.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Directly source or unsource to package manager

This commit migrates the functionality from `AddSource` and
`RemoveSource` from the package manager into the `kraft pkg source` and
`kraft pkg unsource` subcommands, respectively.

This effort redirects the functionality of the package manager's
interface methods of `AddSource` and `RemoveSource` such that these
methods can be used for ephemeral use cases.

Simultaneously, the subcommands `kraft pkg source` and
`kraft pkg unsource` are updated to reflect adding or removing sources
when specifying multiple positional arguments.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Directly handle manifests with option

This commit removes the assumption of populating the list of registries
by utilizing the intended functionality within the option
`WithDefaultRegistries`.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci): Re-work the `IsCompatible` method

This refactor simultaneously solves occassional issue with the method
w.r.t. the determination of a source as an OCI Distribution Spec
compatible registry.  The new implementation uses a slice of functions
to determine whether the provided source is compatible and becomes
easier to understand overall.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Introduce `SetSources` method to `PackageManager`

The method allows for overriding the internal set of sources of the
respective `PackageManager` for the ephemeral usecase.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(oci): Return the image `Manifest` for `Metadata`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(handler): Fix calculation of progress based on host active tasks

In the event that containerd has on-going tasks that are not relevant to
KraftKit and the OCI handler for containerd, this progress calculation
would linger continuously (or until those statuses are ultimately
resolved) therefore representing an inaccurate progress.  If a status
was paused externally, the handler would also linger forever, resulting
in hung progress statuses reported by the handler.

Adjust the `ListStatuses` method to include a filter and check these
per job sequence calculation.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(manifest): Migrate `init` method to own file

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci): Migrate `init` method to own file

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(component): Fix misparsing of a URL of component

This change removes the assumption of a schema prefix of the string
value of the attribute representing the component which would
inaccurately result in mis-version due to the false if-statement.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(run): Introduce runner-led entrypoints for `kraft run`

This commit refactors the `kraft run` subcommand to be more extensible
and to clearly separate the concerns by which the first positional
argument of `kraft run` is used to determine _what_ to run.

In this refactor, the 3 main ways in which a unikernel be run are
clearly separated, both in terms of individual files, but also through
the new abstract interface `runner` which requires implementation to
both determine if possible and then later (if possible) to prepare the
machine interface for the instantiation (population of specification) of
the unikernel VM.  Necessary adjustments to the preparation
(`cmdfactory`'s `Pre` method) and overall body of `cmdfactory`'s `Run`
interface are adjusted, resulting in a larger diff.

The following 3 methods are capable (currently) through the `kraft run`
subcommand:

1. kernel: The first positional argument represents a path to a Unikraft
   unikernel, where the user knows the path of the kernel following a
   build;
2. project: The first positional argument or zero-positional argument
   and current working directory are determined to be a KraftKit-enabled
   repository path (one with a `kraft.yaml`/`Kraftfile` schema); and,
3. package: The first positional argument represents a
   `PackageManager`-compatible package which contains a pre-built
   unikernel.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(unikraft): Introduce miscellaneous Unikraft-centric helper methods

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(elfloader): Introduce ELFLoader Unikraft internal application

This commit introduces an MVP ELFLoader integration into KraftKit which
facilitates the use of the ELFLoader application through the OCI
`PackageManager`.  The constructor method `NewELFLoaderFromPrebuilt`
initially provides the ability to prepare a given `linuxu` binary
through the remotely accessible ELFLoader application via the `Package`
interface.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(run): Introduce Linux Userspace runner to `kraft run`

Using the underlying KraftKit native ELFLoader Unikraft package
implementation, determine whether the provided is a candidate executable
Linux userspace binary and the facilitate its runtime by using the
ELFLoader OCI package.

Signed-off-by: Alexander Jung <alex@unikraft.io>
Co-authored-by: Antoine Cotten <antoine@unikraft.io>

nderjung committed:
feat(v1alpha1): Introduce status path for initramfs

This commit introduces, similar to the kernel path, the status path for
the initram filesystem (CPIO archive).  This is used as status and as
part of the usual instantiation of a machine across existing VMM
solutions.  Simultaneously, the QEMU implementation of the machine
v1alpha1 interface is updataed to accomodate for this status path.
Unlike "rootfs", initramfs is a systematically defined path, whereas
rootfs is inteded to be a user-specified path which should ultimately be
resolved from different high-level locations.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc commented on oci/manager.go:
```diff
@@ -294,133 +292,148 @@ func (manager ociManager) Catalog(ctx context.Context, qopts ...packmanager.Quer
-					Password: auth.Token,
-				})),
-			)
+	// Check if the provided source is  OCI registry
```
```suggestion
	// Check if the provided source is OCI registry
```

craciunoiuc commented on unikraft/utils.go:
```diff
@@ -0,0 +1,48 @@
+		return false, fmt.Errorf("first positional argument is a directory: %v", path)
+	}
+
+	// Sanity check  whether the provided file is an ELF kernel with
```
```suggestion
	// Sanity check whether the provided file is an ELF kernel with
```

craciunoiuc commented on cmd/kraft/run/runner_linuxu.go:
```diff
@@ -0,0 +1,177 @@
+		var v uint64
+		switch ef.Class {
+		case elf.ELFCLASS32:
+			t = elf.DynTag(ef.ByteOrder.Uint32(d[0:4]))
+			v = uint64(ef.ByteOrder.Uint32(d[4:8]))
+			d = d[8:]
+		case elf.ELFCLASS64:
+			t = elf.DynTag(ef.ByteOrder.Uint64(d[0:8]))
+			v = ef.ByteOrder.Uint64(d[8:16])
+			d = d[16:]
```
looks beautiful

nderjung replied:
You can thank @antoineco for this :)

craciunoiuc replied:
it's fine, such is the way of parsing binaries

craciunoiuc reviewed (commented):
doing some brief tests now

nderjung committed:
refactor(run): Add runner aliasing

This commit introduces aliasing to the `PackageManager` to enable users
to specifically set the PM of choice via the `--as` flag.  This keeps
the CLI consistent with other appearances of `--as`.

Simultaneously, the value of the `--as` flag is validated in `Pre`.  We
can unfortunately not use the general-purpose `EnumFlag`, since
injecting the flag in `New` will not resolve registered
`PackageManager`s.  As a result, we introduce a naive alternative.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor: Adjust the paths of the machine state directory

This commit reworks the package runner and the only implementing machine
interface (QEMU) so as to set the state directory of a machine as an
independent directory and not all in the top-level directory
`/var/kraftkit` (the default runtime directory).  This brings
much-needed order to the runtime directory.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc reviewed (approved):
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
